### ISSUE TITLE
Allow the use of a separate file for shading (grdimage -I)

### ISSUE THREAD

**Issue by MarkWieczorek** [GenericMappingTools/pygmt#618]
The gmt command grdimage has an option `-I` that adds shading to another image by modulating the intensity.  Historically, (up until gmt v5), this command has only taken as an option another grd filename. One common use of this feature was to plot something like a gravity, the geoid, or magnetic field in color, and then apply shading using a topography file to provide geologic context.

Since gmt 5, there has been the option of using the input grd file along with grdgradient to automatically generate an intensity file that would be used with the `-I` command. I believe that pygmt only accepts the option `shading=True` which automatically generates such an intensity file using gmt defined default illumination angle parameters.

I am asking that pygmt fully implement the `grdimage -I` option. This is the **only** feature that I am currently missing in pygmt, and which is forcing me to use both gmt and pygmt separately to generate publication quality images for journal articles.

For reference, here is the relevant part of the `gmt grdimage` man page:

```
	-I Apply directional illumination. Append name of intensity grid file.
	   For a constant intensity (i.e., change the ambient light), append a value.
	   To derive intensities from <grd_z> instead, append +a<azim> [-45], +n<method> [t1], and +m<ambient> [0]
	   or use -I+d to accept the default values (see grdgradient for details).
```

**Comment by weiji14** [comment]
Hi @MarkWieczorek, I agree that this will be a useful feature. Do you have an example for this/a map illustration? Even just the pure GMT command will do (best if we can access the data too).

**Comment by MarkWieczorek** [comment]
Well, here is what it should look like (Earth geoid with topography illumination). Sorry for the rainbow colormap, this is a few years old:

![Screen Shot 2020-09-18 at 13 25 57](https://user-images.githubusercontent.com/8697285/93592496-91924a80-f9b2-11ea-8565-21117a929604.png)

I could look through my files, but the scripts will most likely be for gmt 5, and the grd files are pretty big. The only peculiarity with specifying a `-I` file is that the grid size and grid registration need to be exactly the same. If you really need something to test, I could upload them somewhere. But, the easiest test would be to just use the same grid for the colormap as for `-I` and see if you get the same thing as `-I=True`.

**Comment by weiji14** [comment]
I was thinking more along the lines of a [GMT Remote Data-Set](https://docs.generic-mapping-tools.org/latest/datasets/remote-data.html) (i.e. the `@` files), but not sure if there's one that would be suitable? Could you please post the code too for generating this map? Sorry for the fuss, just that I'm not familiar with shading.

**Comment by MarkWieczorek** [comment]
Here is something for a different image

```
grdimage -R-0/360/-90/90 thick.grd -I/Users/lunokhod/Mars/Topography/Mars_16_srm.grd -Csingle.cpt -JW-100/13 -Bg30WSen -Y10 -K -V > output.ps
```
In essence, if you are creating an image with grdimage, you just add the option `-Ifile.grd`

**Comment by weiji14** [comment]
> Here is something for a different image
> 
> ```
> grdimage -R-0/360/-90/90 thick.grd -I/Users/lunokhod/Mars/Topography/Mars_16_srm.grd -Csingle.cpt -JW-100/13 -Bg30WSen -Y10 -K -V > output.ps
> ```
>
> In essence, if you are creating an image with grdimage, you just add the option `-Ifile.grd`

Cool, thanks for the example. This sounds very similar to the `drapegrid`/`-G` option in [`pygmt.grdview`](https://www.pygmt.org/v0.2.0/api/generated/pygmt.Figure.grdview.html), and it could perhaps be implemented similarly:

https://github.com/GenericMappingTools/pygmt/blob/d3e131a863a008f14979e7d9764cc6167156376f/pygmt/base_plotting.py#L433-L444

I'll take a closer look over the weekend (getting late for me here), but you're more than welcome too to open up a Pull Request to speed up the process :smile:.

**Comment by seisman** [comment]
> I believe that pygmt only accepts the option `shading=True` which automatically generates such an intensity file using gmt defined default illumination angle parameters.

I don't test it, but I believe you can already use `shading="/Users/lunokhod/Mars/Topography/Mars_16_srm.grd"` to speify the intensity grid. 

Of course, there are more things to do:

1. the `shading` is not well documented in the `grdimage` function
2. currently we can't pass an xarray grid to `shading`.

**Comment by MarkWieczorek** [comment]
I'll test (1) monday, but (2) is what I really want!

**Comment by seisman** [comment]
Yes, (2) is definitely a missing feature!

**Comment by weiji14** [comment]
> I don't test it, but I believe you can already use `shading="/Users/lunokhod/Mars/Topography/Mars_16_srm.grd"` to speify the intensity grid. Please let us know if it works for you.

I can confirm that this works, here's a very minimal working example:

```python
import pygmt

fig = pygmt.Figure()
fig.grdimage(grid="@earth_relief_01d_g", shading="@earth_relief_01d_g+d")
fig.savefig("shading_with_intensfile.png")
fig.show()
```
![shading_with_intensfile](https://user-images.githubusercontent.com/23487320/93692256-e1088000-fb44-11ea-8314-7fb40958fbec.png)

However, shading an `xarray` grid with an `intensfile` doesn't work, likely due to same issue at #364:

```
fig = pygmt.Figure()
grid = pygmt.datasets.load_earth_relief(registration="gridline")
fig.grdimage(grid=grid, cmap="geo", shading="@earth_relief_01d_g+d")
fig.savefig("xarray_shading_with_intensfile.png")
fig.show()
```
![xarray_shading_with_intensfile](https://user-images.githubusercontent.com/23487320/93692303-7a379680-fb45-11ea-81a7-921bc9e6b14a.png)


> Of course, there are more things to do:
> 
> 1. the `shading` is not well documented in the `grdimage` function
> 
> 2. currently we can't pass an xarray grid to `shading`.

Agree with both points, we can better document `grdimage` and allow for `xarray` grids to `shading` in the same PR. But note that this 2 (shading with xarray grids) probably won't work properly until we fix #364 (an upstream GMT issue).

**Comment by seisman** [comment]
It seems that the upstream GMT PR https://github.com/GenericMappingTools/gmt/pull/4328 also fixes this issue.

Running @weiji14's script above gives me a plot which looks correct (with the GMT master branch `version: 6.2.0_8a4275f_2020.10.14`):

![image](https://user-images.githubusercontent.com/3974108/96063106-2dee2680-0e65-11eb-9de3-725a8fe580a4.png)

But we need some more tests.

**Comment by MarkWieczorek** [comment]
I finally got around to testing this, and grdimage with shading works fine for me as long as I specify a pre-existing filename. For my pipeline, a hack solution is to save an xarray DataArray as a netcdf file, and then to use this filename with the shading option.

Obviously, it would be better if we could just specify the xarray DataArray directly in the shading option.

**Comment by seisman** [comment]
@MarkWieczorek I've implemented the requested feature in #750 and it seems working well (I only did a few simple tests and eye-checking the images). It would be good if you can test it and give some feedbacks.

**Comment by seisman** [comment]
The feature request has been implemented in #750 and now merged in master. Now `grdimage` can accept a netCDF grid or a xarray.DataArray as shading.

However, the grdimage shading with a xarray.DataArray has known upstream bugs (see https://github.com/GenericMappingTools/pygmt/pull/750#discussion_r644475505, https://github.com/GenericMappingTools/pygmt/pull/750#discussion_r647908094, and https://github.com/GenericMappingTools/gmt/issues/5294).

**Comment by weiji14** [comment]
> However, the grdimage shading with a xarray.DataArray has known upstream bugs (see [#750 (comment)](https://github.com/GenericMappingTools/pygmt/pull/750#discussion_r644475505), [#750 (comment)](https://github.com/GenericMappingTools/pygmt/pull/750#discussion_r647908094), and [GenericMappingTools/gmt#5294](https://github.com/GenericMappingTools/gmt/issues/5294)).
> 
> I'm reopening this issue to better track the upstream bug.

The last bits of the upstream CPT stretching issue (https://github.com/GenericMappingTools/gmt/issues/5294) should be fixed with https://github.com/GenericMappingTools/gmt/pull/5947 and https://github.com/GenericMappingTools/gmt/pull/5948 (thanks Meghan and Paul!) and will be available in GMT 6.3.

I've opened a GMT conda-forge dev build at https://github.com/conda-forge/gmt-feedstock/pull/173 which will contain these patches, and allow us to test things easily using `conda install -c conda-forge/label/dev gmt`. So we should be able to close this issue once PyGMT bumps the minimum GMT version to 6.3, and the xfail is removed from the test below:

https://github.com/GenericMappingTools/pygmt/blob/5dc135b3d789c842c959c9abbdb505b07011a68e/pygmt/tests/test_grdimage.py#L120-L123
